server/src/main/java/org/elasticsearch/index/codec/tsdb/ES87TSDBDocValuesEncoder.java

```diff
@@ -187,23 +187,57 @@ void encode(long[] in, DataOutput out) throws IOException {
      *     <li>byte 0: 1/2 bits header+6/7 bits data</li>
      *     <li>byte 1..n: data</li>
      * </ul>
-     * The header (first 1 or 2 bits) describes how the data is encoded:
+     * The header (first 1-3 bits) describes how the data is encoded:
      * <ul>
-     *     <li>?0 block has a single value (vlong), 2nd bit already contains data</li>
+     *     <li><code>?0</code>: block has a single value (vlong), 2nd bit already contains data</li>
      *     <li>
-     *         01 block has two runs, data contains value 1 (vlong), run-length (vint) of value 1,
+     *         <code>01</code>: block has two runs, data contains value 1 (vlong), run-length (vint) of value 1,
      *         and delta from first to second value (zlong)
      *     </li>
-     *     <li>11 block is bit-packed</li>
+     *     <li>
+     *         <code>111</code>: block contains cyclic data, data contains cycle length (vlong),
+     *         and the values until the cycle repeats (encoded as vlongs)
+     *     </li>
+     *     <li><code>011</code>: block is bit-packed</li>
      * </ul>
      */
```

Comment:
What about better formalizing the fact that the encoding strategy is stored as a unary code, ie. the header is a vlong where the number of trailing ones defines the encoding strategy:
 - 0: single run
 - 1: two runs
 - 2: bit-packed
 - 3: cycle

(This number can be computed by doing `int encoding = Long.numberOfTrailingZeros(~v1);`)

(By the way, maybe "bit-packed" should use 7 trailing bits since it doesn't use the higher bits of the first byte anyway, this would allow for future encodings to steal one more bits from the first byte.)

Reply:
Yeah it's not clear to me whether this change has been released already or not. Let's keep it at 2 for now if we're not sure, I'm sure we'll have to create a new version of this doc-values format at some point in the future when we'll have an opportunity to do this.

Reply:
I left the current semantics as-is to not run into these issues. It still allows for extensibility but future encoding schemes will have fewer "free" bits to work with and might need to to occupy another byte for the vlong header.